Schemas.md

```diff
@@ -0,0 +1,521 @@
+
+3. **Define schemas for URL path elements in URL_PATH_ARGS_SCHEMAS**
+
+   * The schemas for URL path elements should be written in URL_PATH_ARGS_SCHEMAS in the handler class.
+   * The keys of URL_PATH_ARGS_SCHEMAS should be the full set of URL path elements and the corresponding values should be the schemas for those args. If there are no URL path elements, then URL_PATH_ARGS_SCHEMAS should be set to {} (an empty dict).
+
+   Example: Let ```exploration_id``` be a data present in the url path. Then, the schema for exploration_id should look like:
+
+   ```python
+   URL_PATH_ARGS_SCHEMAS = {
+       'exploration_id': {
+           'schema': {
+               'type': 'basestring'
+           }
+       }
+   }
+   ```
+
+4. **Define schemas for payload arguments and URL query parameter in HANDLER_ARGS_SCHEMAS**
+
+   * The schemas for payload arguments and URL query parameters are written in HANDLER_ARGS_SCHEMAS in the handler class.
+   * After writing [boilerplate code](#handlers-with-no-arguments) for the HANDLER_ARGS_SCHEMAS, the value corresponding to each request method key (GET/PUT/POST/DELETE) should contain all the payload args and URL query parameters for the corresponding method where each key represents the name of an argument and the corresponding value represents its schema. **Note**: While writing boilerplate code, make sure to remove the request keys which do not correspond to any request method in the handler class.
+
+   Example:  Let ```username``` be an argument passed to the delete request method of a handler class. Then, the schema for the delete request method should look like:
+
+   ```python
+   HANDLER_ARGS_SCHEMAS = {
+       'DELETE': {
+           'username': {
+               'schema': {
+                   'type': 'basestring'
+               }
+           }
+       }
+   }
+   ```
+
+#### Important code pointers
+
+When adding schemas for the args of a particular handler class, some analysis is typically needed. The following points discuss the conventions adopted throughout the codebase for adding schemas to handler classes. **Please read these conventions carefully**.
+
+##### Default and optional arguments
+
+If an argument is not present in a payload/request, and the schema for that argument is defined in the handler, then that argument is treated as “missing”. For missing args, schema utils will raise AssertionError which is represented as InvalidInputException by vaidate_and_normalize_args() method.
+
+To provide default args for a handler, include a key with the name ```default_value``` along with the schema. The value for this key is the default value with which the arg will be updated if no value for that arg is provided in the request. If an argument is optional and it is not supposed to be updated with any default value, then the “default_value” key should contain None.
```

Comment:
The first sentence suggests that the key is only provided when we want to provide default args for a handler. But I think we always want to provide that key (based on the examples below)?

Reply:
I think the first sentence is correct. There are three options for handling missing arguments:

* If there is no `default_value` key, then a missing argument will cause an exception.
* If there is a `default_value` key with the value `None`, then a missing argument will not raise any errors. The normalized data will lack the argument.
* If there is a `default_value` key with a value other than `None` (call the value `x`), then a missing argument will not raise any errors. The normalized data will include the argument with the value `x`.

Here's the relevant code:

https://github.com/oppia/oppia/blob/26b1b5dad4b42a540b2557aba1d27b7bd099f190/core/controllers/payload_validator.py#L58-L67

@vojtechjelinek can you confirm?

Reply:
@Nik-09 Can you confirm Chrises assumption?

Reply:
True, the three points mentioned by @U8NWXD aligns with the idea for writing schema for handler params.
Thank you.